File: controller/rimbot/flight_recorder.py
"""Opt-in bounded native timeline; requests reach durable storage before dispatch."""
import hashlib
import json
import os
from pathlib import Path
import threading
import time
import uuid
from contextlib import contextmanager
from contextvars import ContextVar


_action_context = ContextVar('rimbot_recording_action', default={})


@contextmanager
def recording_action(action_id, goal_id):
    token = _action_context.set(dict(action_id=action_id, goal_id=goal_id) if action_id else {})
    try:
        yield
    finally:
        _action_context.reset(token)


class FlightRecorder:
    def __init__(self, path, *, segment_bytes=8*1024*1024, segments=8, payload_bytes=256*1024):
        if segment_bytes < 1024 or segments < 2 or payload_bytes < 128:
            raise ValueError('Recorder bounds are too small')
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.segment_bytes, self.segments, self.payload_bytes = segment_bytes, segments, payload_bytes
        self.lock = threading.Lock()
        self._stream = None
        self._size = 0
        self.sequence = 0
        self.context = {}
        self.elapsed = 0.0
        self.phase_seconds = dict(lock=0.0, encode=0.0, rotate=0.0, write=0.0, fsync=0.0)
        self.records = 0
        self.truncated = 0
        self.rotations = 0
        self.durable_records = 0
        self.run = os.environ.get('RIMBOT_RUN_ID', uuid.uuid4().hex)
        self.event('coverage', coverage='All BridgeClient.core requests, responses and exceptions, including background reads. Runtime snapshots at persistence boundaries. Requests/errors/outcomes are fsynced; response rows are flushed and become durable at the next durable record or rotation. A host crash can leave an explicit unmatched request. No in-game per-frame/pawn transition trace or model token recording.')

    def event(self, kind, *, context=None, durable=True, **payload):
        began = time.perf_counter()
        with self.lock:
            encoding = time.perf_counter()
            self.phase_seconds['lock'] += encoding-began
            encoded = json.dumps(payload, default=str, separators=(',', ':')).encode()
            if len(encoded) > self.payload_bytes:
                correlation = {key:value for key,value in payload.items()
                               if key in ('request','tool','category') and
                               (isinstance(value,int) or isinstance(value,str) and len(value)<=256)}
                payload = dict(truncated=True, original_bytes=len(encoded), sha256=hashlib.sha256(encoded).hexdigest(),
                               preview=encoded[:self.payload_bytes].decode('utf8', errors='replace'), **correlation)
                self.truncated += 1
                encoded = json.dumps(payload, default=str, separators=(',', ':')).encode()
            self.sequence += 1
            row = dict(version=1, run=self.run, scenario=os.environ.get('RIMBOT_SCENARIO'),
                       sequence=self.sequence, wall_time=time.time(), monotonic=time.monotonic(),
                       kind=kind, context={**(self.context if context is None else context), **_action_context.get()})
            # Reuse the payload encoding used for the size/hash check. Responses
            # can be large; serializing them again adds no evidence.
            header = json.dumps(row, default=str, separators=(',', ':')).encode()
            line = header[:-1]+b',"payload":'+encoded+b'}\n'
            self.phase_seconds['encode'] += time.perf_counter()-encoding
            rotated = time.perf_counter()
            if self._stream is None:
                self._stream = self.path.open('ab')
                self._size = self._stream.tell()
            if self._size >= self.segment_bytes:
                os.fsync(self._stream.fileno())
                self._stream.close()
                self._stream = None
                oldest = self.path.with_name(self.path.name+f'.{self.segments-1}')
                oldest.unlink(missing_ok=True)
                for index in range(self.segments-2, 0, -1):
                    source = self.path.with_name(self.path.name+f'.{index}')
                    if source.exists():
                        source.replace(self.path.with_name(self.path.name+f'.{index+1}'))
                self.path.replace(self.path.with_name(self.path.name+'.1'))
                self.rotations += 1
                self._size = 0
            self.phase_seconds['rotate'] += time.perf_counter()-rotated
            written = time.perf_counter()
            if self._stream is None:
                self._stream = self.path.open('ab')
            self._stream.write(line)
            self._stream.flush()
            self._size += len(line)
            self.phase_seconds['write'] += time.perf_counter()-written
            if durable:
                synced = time.perf_counter()
                os.fsync(self._stream.fileno())
                self.phase_seconds['fsync'] += time.perf_counter()-synced
                self.durable_records += 1
            self.records += 1
            self.elapsed += time.perf_counter()-began
            return self.sequence

    def close(self):
        """Finish the current segment durably; later events can reopen it."""
        with self.lock:
            if self._stream is not None:
                try:
                    self._stream.flush()
                    os.fsync(self._stream.fileno())
                finally:
                    self._stream.close()
                    self._stream = None

    def __del__(self):
        stream = getattr(self, '_stream', None)
        if stream is not None:
            stream.close()

    def stats(self):
        return dict(records=self.records, truncated=self.truncated, rotations=self.rotations,
                    durable_records=self.durable_records,
                    recording_seconds=self.elapsed, retention_segments=self.segments,
                    phase_seconds=dict(self.phase_seconds),
                    segment_bytes=self.segment_bytes, payload_bytes=self.payload_bytes)


_recorder = None


def recorder():
    global _recorder
    path = os.environ.get('RIMBOT_FLIGHT_RECORDER')
    if path and (_recorder is None or _recorder.path != Path(path)):
        if _recorder is not None:
            _recorder.close()
        _recorder = FlightRecorder(path)
    return _recorder if path else None


def read_timeline(path):
    path = Path(path)
    files = sorted((p for p in path.parent.glob(path.name+'.*') if p.suffix[1:].isdigit()),
                   key=lambda p: int(p.suffix[1:]), reverse=True)
    if path.exists():
        files.append(path)
    previous = None
    for file in files:
        for number, line in enumerate(file.read_bytes().splitlines(), 1):
            try:
                row = json.loads(line)
                if (not isinstance(row, dict) or not isinstance(row.get('sequence'), int) or 'kind' not in row
                        or not isinstance(row.get('payload'), dict) or not isinstance(row.get('context'), dict)):
                    raise ValueError('Invalid timeline record')
            except ValueError:
                yield dict(kind='recording_gap', file=str(file), line=number, reason='Incomplete or corrupt record')
                continue
            sequence = row['sequence']
            if previous is None and sequence != 1 or previous is not None and sequence != previous+1:
                yield dict(kind='recording_gap', reason='Retention or sequence discontinuity', before=sequence, after=previous)
            previous = sequence
            yield row

File: go/internal/flightrecorder/flightrecorder.go
// Package flightrecorder is an opt-in bounded native timeline: requests reach
// durable storage before dispatch. It replaces controller/rimgovernor/flight_recorder.py
// for the Go controller. Construction is always explicit; there is no hidden
// global recorder, unlike the Python module's env-var singleton.
package flightrecorder

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

const (
	minSegmentBytes = 1024
	minSegments     = 2
	minPayloadBytes = 128
)

// Recorder appends one JSON line per event to path, rotating into numbered
// segments (path.1 is the newest rotated segment) once the active segment
// reaches SegmentBytes, and retaining at most Segments total files. Every
// exported method is safe for concurrent use.
type Recorder struct {
	path         string
	segmentBytes int64
	segments     int
	payloadBytes int
	run          string

	mu       sync.Mutex
	file     *os.File
	size     int64
	sequence uint64
	records  uint64
	truncate uint64
	rotate   uint64
	durable  uint64
	elapsed  time.Duration
}

// Option configures a Recorder at construction. Defaults match the Python
// FlightRecorder: 8 MiB segments, 8 retained segments, 256 KiB payloads.
type Option func(*Recorder)

func SegmentBytes(n int64) Option { return func(r *Recorder) { r.segmentBytes = n } }
func Segments(n int) Option       { return func(r *Recorder) { r.segments = n } }
func PayloadBytes(n int) Option   { return func(r *Recorder) { r.payloadBytes = n } }

// RunID overrides the recorder's run identifier, otherwise a random hex value.
func RunID(id string) Option { return func(r *Recorder) { r.run = id } }

// New opens (or creates) path for durable append and records one "coverage"
// event describing what the timeline does and does not capture.
func New(path string, opts ...Option) (*Recorder, error) {
	if path == "" {
		return nil, errors.New("flightrecorder: path required")
	}
	r := &Recorder{path: path, segmentBytes: 8 << 20, segments: 8, payloadBytes: 256 << 10}
	for _, opt := range opts {
		opt(r)
	}
	if r.segmentBytes < minSegmentBytes || r.segments < minSegments || r.payloadBytes < minPayloadBytes {
		return nil, errors.New("flightrecorder: recorder bounds are too small")
	}
	if r.run == "" {
		var buf [16]byte
		if _, err := rand.Read(buf[:]); err != nil {
			return nil, fmt.Errorf("flightrecorder: run id: %w", err)
		}
		r.run = hex.EncodeToString(buf[:])
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, fmt.Errorf("flightrecorder: %w", err)
	}
	if _, err := r.Event("coverage", nil, true, map[string]any{
		"coverage": "All bridge.Client native requests, responses and exceptions, including background reads. " +
			"Requests/errors are fsynced; response rows become durable at the next durable record or rotation. " +
			"A host crash can leave an explicit unmatched request. No in-game per-frame/pawn transition trace.",
	}); err != nil {
		return nil, err
	}
	return r, nil
}

// Event appends one record. durable forces an fsync before returning, so a
// caller can be certain the record survives a crash; the natural use is a
// durable "request" row bracketing a non-durable "response" row that becomes
// durable only at the next durable record or segment rotation. context and
// payload are marshaled as JSON objects; a nil map encodes as {}.
func (r *Recorder) Event(kind string, context map[string]any, durable bool, payload map[string]any) (uint64, error) {
	if r == nil {
		return 0, errors.New("flightrecorder: recorder required")
	}
	began := time.Now()
	r.mu.Lock()
	defer r.mu.Unlock()
	if payload == nil {
		payload = map[string]any{}
	}
	encoded, err := json.Marshal(payload)
	if err != nil {
		return 0, fmt.Errorf("flightrecorder: encode payload: %w", err)
	}
	if len(encoded) > r.payloadBytes {
		sum := sha256.Sum256(encoded)
		correlated := map[string]any{
			"truncated":      true,
			"original_bytes": len(encoded),
			"sha256":         hex.EncodeToString(sum[:]),
			"preview":        string(encoded[:r.payloadBytes]),
		}
		for _, key := range []string{"request", "tool", "category"} {
			if value, ok := payload[key]; ok && correlatable(value) {
				correlated[key] = value
			}
		}
		r.truncate++
		encoded, err = json.Marshal(correlated)
		if err != nil {
			return 0, fmt.Errorf("flightrecorder: encode truncated payload: %w", err)
		}
	}
	r.sequence++
	sequence := r.sequence
	if context == nil {
		context = map[string]any{}
	}
	row := struct {
		Version  int             `json:"version"`
		Run      string          `json:"run"`
		Sequence uint64          `json:"sequence"`
		WallTime float64         `json:"wall_time"`
		Kind     string          `json:"kind"`
		Context  map[string]any  `json:"context"`
		Payload  json.RawMessage `json:"payload"`
	}{1, r.run, sequence, float64(began.UnixNano()) / 1e9, kind, context, encoded}
	line, err := json.Marshal(row)
	if err != nil {
		return 0, fmt.Errorf("flightrecorder: encode record: %w", err)
	}
	line = append(line, '\n')
	if err = r.ensureOpen(); err != nil {
		return 0, err
	}
	if r.size >= r.segmentBytes {
		if err = r.rotateLocked(); err != nil {
			return 0, err
		}
	}
	n, err := r.file.Write(line)
	if err != nil {
		return 0, fmt.Errorf("flightrecorder: write: %w", err)
	}
	r.size += int64(n)
	if durable {
		if err = r.file.Sync(); err != nil {
			return 0, fmt.Errorf("flightrecorder: fsync: %w", err)
		}
		r.durable++
	}
	r.records++
	r.elapsed += time.Since(began)
	return sequence, nil
}

func correlatable(value any) bool {
	switch v := value.(type) {
	case string:
		return len(v) <= 256
	case int, int32, int64, uint, uint32, uint64, float64:
		return true
	default:
		return false
	}
}

func (r *Recorder) ensureOpen() error {
	if r.file != nil {
		return nil
	}
	file, err := os.OpenFile(r.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("flightrecorder: open: %w", err)
	}
	info, err := file.Stat()
	if err != nil {
		file.Close()
		return fmt.Errorf("flightrecorder: stat: %w", err)
	}
	r.file = file
	r.size = info.Size()
	return nil
}

// rotateLocked finishes the active segment durably, shifts retained segments
// up by one index (dropping the oldest), and starts a fresh empty segment.
func (r *Recorder) rotateLocked() error {
	if err := r.file.Sync(); err != nil {
		return fmt.Errorf("flightrecorder: rotate fsync: %w", err)
	}
	if err := r.file.Close(); err != nil {
		return fmt.Errorf("flightrecorder: rotate close: %w", err)
	}
	r.file = nil
	oldest := r.segmentPath(r.segments - 1)
	_ = os.Remove(oldest)
	for index := r.segments - 2; index >= 1; index-- {
		source := r.segmentPath(index)
		if _, err := os.Stat(source); err == nil {
			if err = os.Rename(source, r.segmentPath(index+1)); err != nil {
				return fmt.Errorf("flightrecorder: rotate shift: %w", err)
			}
		}
	}
	if err := os.Rename(r.path, r.segmentPath(1)); err != nil {
		return fmt.Errorf("flightrecorder: rotate archive: %w", err)
	}
	r.rotate++
	r.size = 0
	return r.ensureOpen()
}

func (r *Recorder) segmentPath(index int) string { return fmt.Sprintf("%s.%d", r.path, index) }

// Close finishes the current segment durably. Later events reopen it.
func (r *Recorder) Close() error {
	if r == nil {
		return nil
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.file == nil {
		return nil
	}
	syncErr := r.file.Sync()
	closeErr := r.file.Close()
	r.file = nil
	if syncErr != nil {
		return syncErr
	}
	return closeErr
}

// Stats reports counters useful for diagnostics; it never fails.
type Stats struct {
	Records           uint64
	Truncated         uint64
	Rotations         uint64
	DurableRecords    uint64
	RecordingSeconds  float64
	RetentionSegments int
	SegmentBytes      int64
	PayloadBytes      int
}

func (r *Recorder) Stats() Stats {
	r.mu.Lock()
	defer r.mu.Unlock()
	return Stats{
		Records: r.records, Truncated: r.truncate, Rotations: r.rotate, DurableRecords: r.durable,
		RecordingSeconds: r.elapsed.Seconds(), RetentionSegments: r.segments,
		SegmentBytes: r.segmentBytes, PayloadBytes: r.payloadBytes,
	}
}

type actionContextKey struct{}

// WithAction attaches an action/goal correlation pair to ctx; Event calls
// made against a bridge that reads it via ActionFrom include it in context.
func WithAction(ctx context.Context, actionID, goalID string) context.Context {
	if actionID == "" {
		return ctx
	}
	return context.WithValue(ctx, actionContextKey{}, map[string]any{"action_id": actionID, "goal_id": goalID})
}

// ActionFrom reads back the correlation WithAction attached, or nil.
func ActionFrom(ctx context.Context) map[string]any {
	value, _ := ctx.Value(actionContextKey{}).(map[string]any)
	return value
}
